### PR TITLE
CLM support for ASP.NET MVC Framework applications

### PR DESCRIPTION
**Pull request by JcolemanNR**
Adds code level metrics support for ASP.NET MVC Framework applications.

Closes: #1092

### CODE REVIEW

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs [hunk 1]
```diff
@@ -29,17 +30,30 @@ public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall ins
...
+                }
 
-                var transactionName = string.Format("{0}/{1}", controllerName, actionName);
+                string transactionName = controllerName + "/" + actionName;
                 transaction.SetWebTransactionName(WebTransactionType.MVC, transactionName, TransactionNamePriority.FrameworkLow);
 
-                var segment = transaction.StartMethodSegment(instrumentedMethodCall.MethodCall, controllerName, actionName);
+                ISegment segment = transaction.StartMethodSegment(instrumentedMethodCall.MethodCall, controllerName, actionName);
+
+                string fullControllerName = MvcRouteNamingHelper.TryGetControllerFullNameFromObject(controllerContext);
```
Comment (michaelgoin) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs:50]: Minor but by duplicating the `TryGet...` pattern for the full name we are executing some of the same code to get the type where we could try to get the type once and then use the short and long names. Perhaps changing existing usages would be painful, though. I do believe `GetType()` has some magic that makes it pretty fast (and only allocate once if I remember right, the bigger concern) and this should not be a hot path.

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs [hunk 2]
```diff
@@ -29,17 +30,30 @@ public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall ins
...
-                var actionName = MvcRouteNamingHelper.TryGetActionNameFromRouteParameters(instrumentedMethodCall.MethodCall, controllerContext.RouteData);
+                // IMPORTANT: Resist the urge to blindly refactor all of this code to use `var`
+                // IMPORTANT: We are being intentional with types over using `var` here due to
+                // IMPORTANT: the effects of handling a `dynamic` object
+                string controllerName = MvcRouteNamingHelper.TryGetControllerNameFromObject(controllerContext);
+                string actionName = MvcRouteNamingHelper.TryGetActionNameFromRouteParameters(instrumentedMethodCall.MethodCall, controllerContext.RouteData);
 
                 var httpContext = controllerContext.HttpContext;
                 if (httpContext == null)
+                {
```
Comment (michaelgoin) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs:41]: +1

File: tests/Agent/IntegrationTests/IntegrationTests/CodeLevelMetrics/AspNetCoreMvcClmTests.cs [hunk 3]
```diff
@@ -13,11 +13,11 @@
 namespace NewRelic.Agent.IntegrationTests.CodeLevelMetrics
 {
     [NetCoreTest]
-    public class AspNetCoreMvcCodeAttributesTest : NewRelicIntegrationTest<RemoteServiceFixtures.AspNetCoreMvcBasicRequestsFixture>
+    public class AspNetCoreMvcClmTests : NewRelicIntegrationTest<RemoteServiceFixtures.AspNetCoreMvcBasicRequestsFixture>
```
Comment (michaelgoin) [inline tests/Agent/IntegrationTests/IntegrationTests/CodeLevelMetrics/AspNetCoreMvcClmTests.cs:16]: Not sure I agree with the renaming across the test files/classes, given these are already nested in a `CodeLevelMetrics` folder/namespace (Clm = redundant) and these are testing the `code.*` attributes added for code level metrics... but obviously its up to you and team how you want your code organized.

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs [hunk 4]
```diff
@@ -29,17 +30,30 @@ public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall ins
...
                     throw new NullReferenceException("httpContext");
+                }
 
-                var transactionName = string.Format("{0}/{1}", controllerName, actionName);
+                string transactionName = controllerName + "/" + actionName;
```
Comment (michaelgoin) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs:45]: I thought concat of a null string used to throw but looks like I misremembered, so seems fine.

**Review by michaelgoin — APPROVED** [review]
Just some minor comments. Approved.

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs (hunk 4 above)
Comment (JcolemanNR) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs:45]: I didn't think of the null case, but verified it does not crash in a test app.

File: tests/Agent/IntegrationTests/IntegrationTests/CodeLevelMetrics/AspNetCoreMvcClmTests.cs (hunk 3 above)
Comment (JcolemanNR) [reply tests/Agent/IntegrationTests/IntegrationTests/CodeLevelMetrics/AspNetCoreMvcClmTests.cs]: Should be correct now :)

**Comment by JcolemanNR** [comment]
Failing test passes locally, and had never been run in CI before. Merging to feature branch

**Pull request merged**